working/bazel/module-structure.md

```diff
@@ -0,0 +1,234 @@
+
+Based on the above, teams tend to structure their projects _hierarchically_,
+with more specific code living deeper in a sub-package. For example, imagine
+the `ninjacat` project after a few more weeks:
+
+```
+> ninjacat/
+  > app/
+    > views/
+      > checkout/
+        > lib/
+        > test/
+        > BUILD
+      > home/
+        > lib/
+        > test/
+        > BUILD
+  > common/
+    > widgets/
+      > login/
+        > lib/
+        > test/
+        > testing/
+          > lib/
+          > BUILD
+        > BUILD
+```
+
+Already we have the following "packages":
+
+* `package:ninjacat.app.views.checkout`
+* `package:ninjacat.app.views.home`
+* `package:ninjacat.common.widgets.login`
+* `package:ninjacat.common.widgets.login.testing`
+
+It's common to use the [`visibility`][5] property to setup the concept of
+"friend" packages, i.e. packages that are only accessible to _specific_ other
+packages. In the following code, we ensure that the `login` "package" is only
+accessible to packages under `app/`:
+
+```
+# //ninjacat/common/widgets/login/BUILD
+
+dart_library(
+    name = "login",
+    srcs = glob(["lib/**.dart"]),
+    visibility = [
+        "//ninajacat/app:__subpackages__",
+    ],
+)
+```
+
+[5]: https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes
+
+Another common pattern is the concept of `testonly`, and `testing` packages:
+
+```
+# //ninjacat/common/widgets/login/BUILD
+
+dart_test(
+    name = "login_test",
+    srcs = ["test/login_test.dart"],
+    deps = [
+        "//ninjacat/common/widgets/login/testing",
+        "//third_party/dart/test",
+    ],
+)
+```
+
+```
+# //ninjacat/common/widgets/login/testing
+
+dart_library(
+    name = "testing",
+    srcs = glob(["lib/**.dart"]),
+    testonly = 1,
+    deps = [
+        "//third_party/dart/pageloader",
+    ],
+)
+```
+
+In this case, `package:ninjacat.common.widgets.login.testing` exposes a set of
+test-only libraries that can be used, only in _test_ targets. This pattern isn't
+replicable externally at all, but is _very_ common internally to avoid bringing
+in test utilities and code into production applications.
+
+## Notable Differences
+
+### No cyclical _targets_
+
+Dart, as the language, allows cyclical dependencies between `.dart` files 
+libraries). Bazel does _not_ allow cyclical dependencies between packages
+(i.e. _targets_). So, the following is illegal in Bazel where it is fine
+externally with `pub`:
+
+```
+# //ninjacat/common/foo (i.e. package:ninjacat.common.foo/foo.dart)
+
+dart_library(
+    name = "foo",
+    srcs = ["lib/foo.dart"],
+    dependencies = [
+        "//ninjacat/common/bar",
+    ],
+)
+```
+
+```
+# //ninjacat/common/bar (i.e. package:ninjacat.common.bar/bar.dart)
+
+dart_library(
+    name = "bar",
+    srcs = ["lib/bar.dart"],
+    dependencies = [
+        "//ninjacat/common/foo",
+    ],
+)
+```
+
+Unfortunately this is quite common when you take into account the concept of
```

Comment:
It's only unfortunate if you want to treat an entire pub package as a single unit of compilation/modularity, which was never the goal.

A pub package's granularity is the unit of "shippable API product". It's a version manager, not a build tool. Finding a pub package lacking as the unit of compilation is like complaining that you can't recline an entire car. That's not the point. It's a container around things that can be reclined. :)

Reply:
But you can't say micro packages are an anti-pattern _and_ say `pubspec.yaml` is only about version solving :-/

Reply:
I think what we probably want are three levels of granularity:

1. **Shippable versioned API product.** These have names, release schedules, dependencies on other products. Cycles are allowed. Pub packages.

2. **Build target.** This is the unit of compilation and modularity. Likely the unit of encapsulation for things like "sealed". Possibly having a privacy control at this level. Acyclic. This doesn't currently exist in a user-facing way.

3. **Source file.** A single file of source code with its own top-level scope. Privacy and namespacing works at this level. Dart libraries. Cycles are allowed (and heavily used in practice).

Reply:
The levels of granularity you describe @munificent are exactly what we have in bazel, with the only difference being that packages are defined as directories with BUILD files, instead of coming from pub. But conceptually they are the same.

What @matanlurey and probably most people who have tried to integrate with a Dart compiler wants/needs, is a formal and consistent way of defining the **Build target** portion externally. I believe this should be integrated with pub, for a couple reasons:

- You need to have separate dependencies per build target in order to efficiently create an acyclic graph of build targets. 
  - The only alternatives I can come up with are constantly re-analyzing your actual Dart import structure to create custom targets (what package:build does, at significant expense), or forcing all end users to create static configuration for all their _transitive_ dependencies, which is just not realistic.
- Allows the build target structure to be statically configured, which is a huge win for build systems. Neither bazel nor package:build_runner allow you to read files before declaring your inputs/outputs, so this is a huge pain point today.
- Module granularity/structure is consistent across tools.
- No extra dependencies for platforms/features you don't use.
- Bazel has proven that configuring build targets in the package itself works well (enough). If you need a more granular target you can always send a PR to a package to create it. See my next point for where this potentially falls down externally though...

All this being said there is one potentially  _huge_ issue with trying to push the bazel model on the external world, which is figuring out how to actually enforce the acyclic requirement of the build targets. With bazel and a single version restriction for packages, this is not really an issue. But for pub packages any new non-breaking release of a package could introduce a build target cycle for any package that depends on that package accidentally. Maybe the version solver could take that into account (throw out versions if they introduce a build target cycle?), but it doesn't seem trivial.

Reply:
Correct. (We should probably add a doc about build system constraints somewhere)